### PR TITLE
Change `ClientApi` to be a facade in front of the store

### PR DESCRIPTION
**Pull request by tmeasday**
Part of #15871

Telescoping on #16005

## What I did

Change the behaviour of the v6 APIs/entrypoints to use the on-demand store in a pre-loaded (i.e. not on-demand) way.

- Change the `@storybook/client-api` API exports (which are re-exported by the various app layers) to be a "facade" in front of the story store. What this means is:
  - all calls to `addDecorators()` & friends are collected on the Client API, and passed to a `WebPreview` as a `getProjectAnnotations` function.
  - all calls to `storiesOf()` are are collected, and massaged into a "CSF import" that is then supplied as a (synchronous) `importFn`
  - all calls to `.add()` are similarly added to the CSF imports, and also indexed into to be passed as a (synchronous) `fetchStoryIndex` function. **Note this function implements story sorting in the preview[1]**

- Similarly, if a `Loadable` is passed to the `configure()` function (like a `require.context`), its exports (CSF exports) are collected by the `ClientApi` in a similar way to the above.

- When the `start()` function (called by all e.g. `@storybook/react` entry points) is called it creates the above and a webpreview, which it `initialize()`s synchronously -- which means all stories are loaded right away.

[1] In the v7 store the story sorting of the index happens in node.

### CODE REVIEW

**Review by shilman — COMMENTED** [review]
We still need to add story sorting to the index server, right?

File: lib/client-api/src/ClientApi.test.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  addons.setChannel(mockChannel());
+});
+
+describe('ClientApi', () => {
+  describe('setAddon', () => {
+    it('should register addons', () => {
+      const clientApi = new ClientApi();
+      let data;
+
+      clientApi.setAddon({
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply lib/client-api/src/ClientApi.test.ts:14]: Good idea, want to open a ticket for it?

**Comment by tmeasday** [comment]
> We still need to add story sorting to the index server, right?

Yes we do.

File: lib/client-api/src/ClientApi.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shilman) [reply lib/client-api/src/ClientApi.test.ts:14]: https://github.com/storybookjs/storybook/issues/16007

**Comment by bebraw** [comment]
> Yes we do. I think @bebraw has prototyped it over in the skeleton.

Yup, I have a basic implementation in place there at https://github.com/tmeasday/storybook-skeleton/pull/28 .

**Comment by shilman** [comment]
Merged as part of https://github.com/storybookjs/storybook/pull/15871

**Pull request closed without merging**